andrepereiradasilva/joomla-cms
# [menu item] Improve alias process

#### Summary of Changes

This PR makes several changes on the menu alias generate process.
- Improves alias generation
- Improves alias conflict detection
- Improves the information in the warning messages so it's easier to find the duplicate alias.
- Always save

Also soves some issues in the current alias generate process:
1. In a monolingual site (language filter plugin disabled), create a menu item with the same alias as other menu item you have in the same/other menu root but with different language.
   
   You'll notice it will save with success but it shouldn't.
2. In a multilingual site (language filter plugin enabled), create a menu item with the same alias as other menu item with All language you have in the same/other menu root but in a different language.
   
   You'll notice it will save with success but it shouldn't.
3. Create a menu item of System Links -> _Any type_
   
   You'll notice the generate alias is the current date, not menu item title.
4. Create a menu item with the same menu alias and same language as other item you have in the menu root.
   
   You'll get a message and will not save.
   ![image](https://cloud.githubusercontent.com/assets/9630530/15653061/80de0ea2-2682-11e6-93bb-c9c73c0987a2.png)
#### Testing Instructions
- Install one adicional language (example: fr-FR)
- Create the following menu structure:

```
 Menu Type 1:
 - Menu item A: Language All, Alias: menu-item-a
 - Menu item B: Language en-GB, Alias: menu-item-b
 - Menu item C: Language fr-FR, Alias: menu-item-c
 Menu Type 2:
 - Menu item D: Language All, Alias: menu-item-d
 - Menu item E: Language en-GB, Alias: menu-item-e
 - Menu item F: Language fr-FR, Alias: menu-item-f
```
- Apply patch
- Set the site to monolingual (Disable Language Filter System plugin)
- Change the alias in `Menu item B` to `menu-item-a`. It should save it with alias `menu-item-a-2` and give a warning message.
  ![image](https://cloud.githubusercontent.com/assets/9630530/15654440/21ed129e-268c-11e6-85ad-75a686331dcf.png)
- Change the alias in `Menu item F` to `menu-item-a`. It should save with alias `menu-item-a-3` and give a warning message.
  ![image](https://cloud.githubusercontent.com/assets/9630530/15654490/9a5a74ba-268c-11e6-8589-945e2ac305e1.png)
- Now set the site to multilingual (Enable Language Filter System plugin)
- Change the alias in `Menu item B` to `menu-item-a`. It should save with alias `menu-item-a-2` and give a warning message.
  ![image](https://cloud.githubusercontent.com/assets/9630530/15654454/4ea6a2d2-268c-11e6-8770-79bbeea04392.png)
- Change the alias in `Menu item F` to `menu-item-e`. It should save fine.
- Change the alias in `Menu item E` to `menu-item-a`. It should save with alias `menu-item-a-3` and give a warning message.
  ![image](https://cloud.githubusercontent.com/assets/9630530/15654468/781174f8-268c-11e6-9332-ab11aad9dc72.png)
- Now create a menu item of System Type -> Menu alias and let it auto generate the alias. It should generate fine.
- Play a little with the alias and check everything is fine (test with and without unicode in global config).
#### Observations

The alias cannot be duplicated in the following cases:
###### Monolingual Site (language filter plugin disabled)
1. The menu item is at root level and has the same alias of other menu item that is at root level of any other menu.
2. The menu item is not at root level, but was the has the same alias of other menu item that has the  same parent.
###### Multilingual Site (language filter plugin enabled)
1. The menu item language is "All", is at root level and has the same alias of other menu item (any language) that is at root level of any other menu.
2. The menu item language is NOT "All", is at root level and has the same alias of other menu item (same or "All" language) that is at root level of any other menu.
3. The menu item language is "All", is not at root level, but was the has the same alias of other menu item (any language) that has the  same parent.
4. The menu item language is "All", is not at root level, but was the has the same alias of other menu item (same or "All" language) that has the  same parent.
